### DISCUSSION TITLE
Astro global to determine output mode

### DISCUSSION THREAD

**Discussion by targetlucked69** [withastro/roadmap#1002] [Proposal]
### Body

# Summary

A new global attached to `Astro` to get the current output/prerender mode of a page in any Astro component.

# Background & Motivation

In Astro, components can be rendered differently depending on whether the application is running in `static`, `server`, or `hybrid` mode. Currently, there's no straightforward way for developers to determine the current rendering mode within a component. This can lead to challenges when trying to conditionally render components or access server-only features like `Astro.locals`.

It's worth noting that this functionality could potentially be implemented using a Vite plugin that checks the Astro configuration to determine if the output is set to `static` or `server`. Example:

```ts
import type { AstroConfig } from 'astro';

export function vitePluginOutputConfig(config: AstroConfig) {
  const virtualModuleId = 'virtual:output-config;
  const resolvedVirtualModuleId = '\0' + virtualModuleId;

  return {
    name: 'vite-plugin-output-config',
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return `
          export const output = "${config.output}"
        `;
      }
    },
  };
}
```

However, this approach becomes challenging when the output is set to `hybrid`, as the rendering mode can vary on a per-page basis.

# Goals

- Provide a simple, intuitive way for developers to check the current output mode (static or server) within Astro components.
- Enable conditional rendering based on the output mode without relying on external configuration or build-time constants.
- Improve the developer experience when creating components that need to work in both static and server environments.
- Accurately represent the rendering mode even in hybrid output scenarios, where the mode may vary per page.

# Example

This RFC proposes adding a new `Astro.output` global that will have a value of either `'static'` or `'server'`. Here's an example of how it could be used:

```astro
---
const Component = Astro.output === 'static' ? StaticComponent : ServerComponent;
---

<Component>
  <slot />
</Component>
```

In this example, `ServerComponent` might rely on `Astro.locals.user`, which is not available in static mode. With `Astro.output`, developers can easily choose the appropriate component based on the current rendering mode.

This new global would provide a clean and efficient way to handle mode-specific logic within Astro components, improving code readability and maintainability. It would be particularly useful in hybrid output scenarios, where a simple Vite plugin checking the Astro config would not suffice to determine the rendering mode of individual pages.

**Comment by Fryuni** [comment]
(self-plug 😜) You can use Astro When to differentiate it when your code is running: https://inox-tools.fryuni.dev/astro-when

> It would be particularly useful in hybrid output scenarios, where a simple Vite plugin checking the Astro config would not suffice to determine the rendering mode of individual pages.

It does, that's how Astro When is implemented.

> However, this approach becomes challenging when the output is set to hybrid, as the rendering mode can vary on a per-page basis.

The output mode doesn't vary on a per-page basis. If it is `hybrid` it will be hybrid everywhere. The difference between `hybrid` and `server` is just a default value, everything else is the same.

```astro
---
// Always prerender, regardless of `hybrid` or `server`
export const pretender = true;
---
```
```astro
---
// Never prerender, regardless of `hybrid` or `server`
export const pretender = false;
---
```
```astro
---
// Use the default prerendering option.
// On `hybrid`, prerender by default.
// On `server`, do not prerender by default.
---
```

You can have pre-rendered pages with `output: 'server'`

**Comment by florian-lefebvre** [comment]
FYI in 5.0 I'm going to work on this, that will be `context.prerender` most likely

**Reply by targetlucked69** [reply]
In Vite land, how will you be able to check if a component for example, is imported in a static/server page with this?

**Reply by florian-lefebvre** [reply]
This is not intended for vite plugins, only for usage in astro context (middleware, .astro files, endpoints). What's your usecase?

EDIT: maybe you can be interested in #998
